substrate/primitives/state-machine/src/overlayed_changes/changeset.rs

```diff
@@ -204,10 +291,208 @@ impl<V> OverlayedEntry<V> {
 	}
 }
 
-impl OverlayedEntry<Option<StorageValue>> {
+/// When a transaction layer is dropped, pass the current data buffer to the
+/// parent layer (will be new current).
+fn restore_append_to_parent(
+	parent: &mut StorageEntry,
+	mut current_data: Vec<u8>,
+	current_materialized: Option<u32>,
+) {
+	match parent {
+		StorageEntry::Append {
+			data: parent_data,
+			nb_append: _,
+			materialized: parent_materialized,
+			from_parent: _,
+		} => {
+			let AppendData::MovedSize(mut target_size) = parent_data else {
+				unreachable!("restore only when parent is moved");
```

Comment:
Would it be possibly to return an error here?

Reply:
That's a bit awkward (all overlay is written without error return and with a lot of expect), but I will try (end of story it will run expect at host function level as they don't return an error, but may be cleaner).

Reply:
I did try https://github.com/paritytech/polkadot-sdk/pull/1223/commits/bd3c28e8679802c680482da41b8bee3cb72a66a2 but revert it as I find it too awkward.
Issue here is that the sequence of item in the transaction is supposed to be in a certain way only (MovedSize follow at the very end by Data). Similarily some state are only possible if there is a parent (from_parent being true), but then it also lead to this kind of expects.
I don't really know if I can find a way to force these from the data structure (data can be followed by a removal so it is not only in head, but movedsize is always followed by data).